backend/de.metas.banking/de.metas.banking.base/src/main/java/de/metas/banking/payment/impl/PaySelectionBL.java

```diff
@@ -71,51 +84,50 @@ public void updateFromInvoice(final I_C_PaySelectionLine psl)
-		}
-
-		final IBPBankAccountDAO bpBankAccountDAO = Services.get(IBPBankAccountDAO.class);
+		final InvoiceId invoiceId = InvoiceId.ofRepoIdOrNull(psl.getC_Invoice_ID());
+		if (invoiceId == null) {return;}
+		final I_C_Invoice invoice = invoiceBL.getById(invoiceId);
 
 		final Properties ctx = InterfaceWrapperHelper.getCtx(psl);
 
-		final int partnerID = invoice.getC_BPartner_ID();
-		psl.setC_BPartner_ID(partnerID);
+		final BPartnerId partnerID = BPartnerId.ofRepoId(invoice.getC_BPartner_ID());
+		psl.setC_BPartner_ID(partnerID.getRepoId());
 
 		// task 09500 get the currency from the account of the selection header
 		// this is safe because the columns are mandatory
-		final int currencyID = psl.getC_PaySelection().getC_BP_BankAccount().getC_Currency_ID();
-
-		final boolean isSalesInvoice = invoice.isSOTrx();
+		final I_C_PaySelection paySelection = Check.assumePresent(paySelectionDAO.getById(PaySelectionId.ofRepoId(psl.getC_PaySelection_ID())),"Pay Selection should be present");
+		final BankAccount bankAccount = bpBankAccountDAO.getById(BankAccountId.ofRepoId(paySelection.getC_BP_BankAccount_ID()));
+		final CurrencyId currencyID = bankAccount.getCurrencyId();
 
-		final IInvoiceBL invoiceBL = Services.get(IInvoiceBL.class);
-		final boolean isCreditMemo = invoiceBL.isCreditMemo(invoice);
+		psl.setC_BP_BankAccount_ID(BankAccountId.toRepoId(getBankAccountId(invoiceId, currencyID, ctx)));
 
-		final String accteptedBankAccountUsage;
-
-		if ((isSalesInvoice && !isCreditMemo) ||
-				(!isSalesInvoice && isCreditMemo))
-		{
-			// allow a direct debit account if there is an invoice with SOTrx='Y', and not a credit memo
-			// OR it is a Credit memo with isSoTrx = 'N'
-			accteptedBankAccountUsage = X_C_BP_BankAccount.BPBANKACCTUSE_DirectDebit;
-		}
-		else
+		if (Check.isBlank(psl.getReference()) && InterfaceWrapperHelper.isNew(psl))
 		{
-			// allow a direct deposit account if there is an invoice with SOTrx='N', and not a credit memo
-			// OR it is a Credit memo with isSoTrx = 'Y'
-			accteptedBankAccountUsage = X_C_BP_BankAccount.BPBANKACCTUSE_DirectDeposit;
+			psl.setReference(invoice.getPOReference());
 		}
+	}
 
-		final List<I_C_BP_BankAccount> bankAccts = bpBankAccountDAO.retrieveBankAccountsForPartnerAndCurrency(ctx, partnerID, currencyID);
+	@Nullable
+	@Override
+	public BankAccountId getBankAccountId(@NonNull final InvoiceId invoiceId,
+										  @NonNull final CurrencyId currencyId,
+										  @NonNull final Properties ctx)
+	{
+		final I_C_Invoice invoice = invoiceBL.getById(invoiceId);
+		final boolean isSalesInvoice = invoice.isSOTrx();
+		final boolean isCreditMemo = invoiceBL.isCreditMemo(invoice);
+
+		final String accteptedBankAccountUsage = getAcceptedBankAccountUsage(isSalesInvoice, isCreditMemo);
+
+		final List<I_C_BP_BankAccount> bankAccts = bpBankAccountDAO.retrieveBankAccountsForPartnerAndCurrency(
```

Comment:
suggestion: would be to much work if we make retrieveBankAccountsForPartnerAndCurrency return POJOs instead of I_C_....?